File: examples/simple_robot/src/main/java/frc/robot/subsystems/TurretSubsystem.java
package frc.robot.subsystems;


import static edu.wpi.first.units.Units.Amps;
import static edu.wpi.first.units.Units.Degrees;
import static edu.wpi.first.units.Units.Feet;
import static edu.wpi.first.units.Units.Radians;
import static edu.wpi.first.units.Units.RadiansPerSecond;
import static edu.wpi.first.units.Units.Second;
import static edu.wpi.first.units.Units.Volts;

import com.ctre.phoenix6.hardware.TalonFX;
import edu.wpi.first.math.controller.ArmFeedforward;
import edu.wpi.first.math.geometry.Pose2d;
import edu.wpi.first.math.geometry.Rotation2d;
import edu.wpi.first.math.geometry.Rotation3d;
import edu.wpi.first.math.geometry.Transform2d;
import edu.wpi.first.math.geometry.Transform3d;
import edu.wpi.first.math.geometry.Translation2d;
import edu.wpi.first.math.kinematics.ChassisSpeeds;
import edu.wpi.first.math.system.plant.DCMotor;
import edu.wpi.first.units.measure.Angle;
import edu.wpi.first.wpilibj2.command.Command;
import edu.wpi.first.wpilibj2.command.SubsystemBase;
import yams.gearing.GearBox;
import yams.gearing.MechanismGearing;
import yams.mechanisms.config.PivotConfig;
import yams.mechanisms.positional.Pivot;
import yams.motorcontrollers.SmartMotorController;
import yams.motorcontrollers.SmartMotorControllerConfig;
import yams.motorcontrollers.SmartMotorControllerConfig.ControlMode;
import yams.motorcontrollers.SmartMotorControllerConfig.MotorMode;
import yams.motorcontrollers.SmartMotorControllerConfig.TelemetryVerbosity;
import yams.motorcontrollers.remote.TalonFXWrapper;

public class TurretSubsystem extends SubsystemBase
{

  double[] ratio = {144 / 15, 5, 1.08};

  SmartMotorControllerConfig motorConfig = new SmartMotorControllerConfig(this)
      .withControlMode(ControlMode.CLOSED_LOOP)
      .withSimClosedLoopController(0.0, 0.0, 0)
      // 99.0, 0.0, .6
      .withClosedLoopController(0.0, 0.0, 0)

      // Configure Motor and Mechanism properties
      .withGearing(new MechanismGearing(new GearBox(ratio)))
      .withIdleMode(MotorMode.BRAKE)
      .withMotorInverted(false)
      .withFeedforward(new ArmFeedforward(0.5, 0.0, 5.0, 0))
      .withSimFeedforward(new ArmFeedforward(0.5, 0.0, 5.0, 0))

      // 0.0,5.5`
      // Setup Telemetry
      .withTelemetry("TurretMotor", TelemetryVerbosity.HIGH)
      // Power Optimization
      .withStatorCurrentLimit(Amps.of(60));
  // .withClosedLoopRampRate(Seconds.of(0.0))

  // .withOpenLoopRampRate(Seconds.of(0.0));
  SmartMotorController motor = new TalonFXWrapper(new TalonFX(12),
                                                  DCMotor.getKrakenX60(1),
                                                  motorConfig);

  PivotConfig m_config = new PivotConfig(motor)
      .withStartingPosition(Degrees.of(0)) // Starting position of the Pivot
      .withHardLimit(Degrees.of(-360), Degrees.of(360)) // Hard limit bc wiring prevents infinitpe spinning
      // .withSoftLimits(Degrees.of(-360), Degrees.of(360))
      .withTelemetry("Turret", TelemetryVerbosity.HIGH) // Telemetry
      .withMOI(yams.units.YUnits.PoundSquareInches.of(0.01)); // MOI Calculation

  private Pivot turret = new Pivot(m_config);


  // Robot to turret transform, from center of robot to turret.
  private final Transform3d roboToTurret = new Transform3d(Feet.of(-1.5), Feet.of(0), Feet.of(0.5), Rotation3d.kZero);

  public TurretSubsystem()
  {
    // TODO: Set the default command, if any, for this subsystem by calling setDefaultCommand(command)
    //       in the constructor or in the robot coordination class, such as RobotContainer.
    //       Also, you can call addChild(name, sendableChild) to associate sendables with the subsystem
    //       such as SpeedControllers, Encoders, DigitalInputs, etc.
  }

  public Pose2d getPose(Pose2d robotPose)
  {
    return robotPose.plus(new Transform2d(
        roboToTurret.getTranslation().toTranslation2d(), roboToTurret.getRotation().toRotation2d()));
  }

 public ChassisSpeeds getVelocity(ChassisSpeeds robotVelocity, Angle robotAngle)
  {

      Translation2d rRobot = roboToTurret.getTranslation().toTranslation2d(); // in robot frame
    Translation2d rWorld = rRobot.rotateBy(Rotation2d.fromRadians(robotAngle.in(Radians))); // rotate into field frame

      double omega = robotVelocity.omegaRadiansPerSecond; // robot yaw rate (rad/s)

      // rotational linear velocity at turret (v_rot = ω × r_world)
      double vRotX = -omega * rWorld.getY();
      double vRotY =  omega * rWorld.getX();

      // final turret linear velocity in field frame
      double turretVx = robotVelocity.vxMetersPerSecond + vRotX;
      double turretVy = robotVelocity.vyMetersPerSecond + vRotY;

      // turret angular velocity in field frame
      double turretOmega = omega + motor.getMechanismVelocity().in(RadiansPerSecond);

      return new ChassisSpeeds(turretVx, turretVy, turretOmega);

  }

  public void periodic()
  {
    turret.updateTelemetry();
  }

  public void simulationPeriodic()
  {
    turret.simIterate();
  }

  public Command turretCmd(double dutycycle)
  {
    return turret.set(dutycycle);
  }

  public Command sysId()
  {
    return turret.sysId(Volts.of(3), Volts.of(3).per(Second), Second.of(30));
  }

  public Command setAngle(Angle angle)
  {
    return turret.setAngle(angle);
  }

  public void setAngleSetpoint(Angle measure)
  {
    turret.setMechanismPositionSetpoint(measure);
  }
}

File: examples/cpptest/src/main/cpp/subsystems/TurretSubsystem.cpp
// Copyright (c) 2026 Yet Another Software Suite
// SPDX-License-Identifier: LGPL-3.0-or-later

// Turret pivot (yaw-only) driven by a Kraken X60 (CAN 12) through a three-stage compound
// gearbox (~51.84:1). Mounted 1.5 ft behind and 0.5 ft above the robot center.
// GetPose / GetVelocity translate that offset into field-relative coordinates for targeting.

#include "subsystems/TurretSubsystem.h"

#include <frc/geometry/Rotation2d.h>
#include <frc/geometry/Transform2d.h>
#include <frc/geometry/Translation2d.h>
#include <frc/system/plant/DCMotor.h>
#include <units/angle.h>
#include <units/angular_velocity.h>
#include <units/current.h>
#include <units/length.h>

#include <cmath>
#include <iostream>
#include <numbers>

using namespace yams::motorcontrollers;
using namespace yams::gearing;
using namespace yams::mechanisms;
using Cfg = SmartMotorControllerConfig;

TurretSubsystem::TurretSubsystem()
    // Pivot is 1.5 ft behind (-X) and 0.5 ft above (+Z) the robot origin; no yaw offset
    // at construction -- the turret's angular setpoint is relative to the robot heading.
    : m_roboToTurret{frc::Translation3d{units::foot_t{-1.5}, units::foot_t{0}, units::foot_t{0.5}},
                     frc::Rotation3d{}} {
  m_motorConfig.WithSubsystem(this)
      .WithClosedLoopMode()
      // kP=0 placeholder -- the turret is not yet PID-tuned. Set kP first; add kD for
      // damping if the turret oscillates when tracking a moving target.
      .WithFeedback(0.0, 0.0, 0)
      // Three-stage gearbox expressed as individual reduction factors, multiplied together:
      //   stage 1: 144/15 = 9.6:1, stage 2: 5:1, stage 3: 1.08:1 -> 51.84:1 total.
      // The initializer-list GearBox constructor multiplies all stages, same as
      // FromReductionStages.
      .WithMotorGearing(MechanismGearing{GearBox({144.0 / 15.0, 5.0, 1.08})})
      // BRAKE prevents the turret from drifting off-target when no command is running.
      // On a turret with a heavy mechanism attached, COAST can cause dangerous snap rotation.
      .WithIdleMode(Cfg::MotorMode::BRAKE)
      .WithMotorInverted(false)
      // WithArmFeedforward on a turret: kS=0.5 V overcomes static friction at the worm/gear
      // interface; kV=5.0 V/(turn/s) is the velocity gain for tracking; kA=0 (no accel FF yet);
      // kG=0 because the turret rotates in the horizontal plane -- gravity does no work.
      .WithArmFeedforward(0.5, 5.0, 0, 0.0)
      // Turret faces forward (0 deg = robot heading) at power-on. If the turret has a
      // hard-stop home position, replace 0 with the measured home angle.
      .WithStartingPosition(units::degree_t{0})
      .WithTelemetry("TurretMotor", Cfg::TelemetryVerbosity::HIGH)
      // Stator limit (output side): 60 A caps peak torque on the output shaft.
      // Using stator (not supply) here because we care about peak force, not battery draw.
      .WithStatorCurrentLimit(units::ampere_t{60});

  // KrakenX60(1): single-motor sim model for the turret. The (1) is the motor count
  // used to average free-speed and stall torque -- leave it at 1 for a single-motor drive.
  m_motor.emplace(&m_talonFX, frc::DCMotor::KrakenX60(1), &m_motorConfig);

  // Soft limits span a full revolution in each direction. This allows 360-degree tracking
  // without a hard stop but prevents winding the cable harness past two full turns.
  m_pivotConfig.WithMinAngle(units::degree_t{-360})
      .WithMaxAngle(units::degree_t{360})
      .WithTelemetryName("Turret");

  m_turret.emplace(&m_pivotConfig, &m_motor.value());
}

frc::Pose2d TurretSubsystem::GetPose(frc::Pose2d robotPose) const {
  // Projects the robot-to-turret 3D transform down to 2D for use with field-relative
  // pose estimators. The Z component (height) is discarded; only XY offset and yaw matter.
  return robotPose.TransformBy(frc::Transform2d{m_roboToTurret.Translation().ToTranslation2d(),
                                                m_roboToTurret.Rotation().ToRotation2d()});
}

frc::ChassisSpeeds TurretSubsystem::GetVelocity(frc::ChassisSpeeds robotVelocity,
                                                units::degree_t robotAngle) {
  // Compute the linear velocity at the turret pivot due to the robot's own rotation.
  // v = omega x r: in 2D, vx = -omega * ry, vy = omega * rx (right-hand cross product).
  // rWorld is the pivot offset rotated to the field frame so the cross product is correct.
  auto rRobot = m_roboToTurret.Translation().ToTranslation2d();
  auto rWorld = rRobot.RotateBy(frc::Rotation2d{robotAngle});

  double omega = robotVelocity.omega.value();  // rad/s

  double vRotX = -omega * rWorld.Y().value();
  double vRotY = omega * rWorld.X().value();

  units::meters_per_second_t turretVx{robotVelocity.vx.value() + vRotX};
  units::meters_per_second_t turretVy{robotVelocity.vy.value() + vRotY};

  // Add the turret's own rotational velocity (deg/s -> rad/s) to the robot yaw rate
  // so turret-relative omega is correct for a shooter that follows the turret barrel.
  double mecVelDegPerS = m_motor->GetMechanismVelocity().value();
  double mecVelRadPerS = mecVelDegPerS * std::numbers::pi / 180.0;
  units::radians_per_second_t turretOmega{omega + mecVelRadPerS};

  return frc::ChassisSpeeds{turretVx, turretVy, turretOmega};
}

void TurretSubsystem::SetAngleSetpoint(units::degree_t angle) {
  // Direct setpoint write -- useful for auto-aim from Periodic without wrapping in a command.
  // If a SetAngle() command is also running, its next iteration will overwrite this.
  m_turret->SetMechanismPositionSetpoint(angle);
}

void TurretSubsystem::Periodic() { m_turret->UpdateTelemetry(); }

void TurretSubsystem::SimulationPeriodic() { m_turret->SimIterate(); }

frc2::CommandPtr TurretSubsystem::TurretCmd(double dutycycle) {
  // Bypasses the Pivot mechanism and writes duty-cycle directly to the motor wrapper.
  // The "HIIII" cout is debug scaffolding -- remove before competition.
  return Run([this, dutycycle] {
    m_motor->SetDutyCycle(dutycycle);
    std::cout << "HIIII";
  });
}

frc2::CommandPtr TurretSubsystem::SetAngle(units::degree_t angle) { return m_turret->Run(angle); }
